File: Sources/MurmurDictionary/DictionaryCorrector.swift
import Foundation

/// One correction that actually fired, kept so history can show whether the dictionary is
/// earning its place.
public struct AppliedCorrection: Codable, Hashable, Sendable {
    /// The text as the engine produced it.
    public let from: String
    /// What it was rewritten to.
    public let to: String
    /// How many times it fired in this transcript.
    public let count: Int
}

/// Rewrites transcribed text using the dictionary's correction pairs.
///
/// This is the guaranteed half of the dictionary. Engine biasing is a nudge — it raises the
/// odds of the right word and promises nothing — so anything that must be correct has to be
/// fixed here, after the fact, deterministically.
///
/// Two passes, in order. **The literal pass** applies the correction pairs you typed out.
/// **The phonetic pass** then catches spans that merely *sound* like something the dictionary
/// knows, which is what stops a name needing one rule per mishearing.
///
/// Three rules govern the literal pass, all load-bearing:
///
/// **Longest match first.** "Claude Code" is applied before "Claude", so the longer rule
/// isn't pre-empted by a shorter one that overlaps it.
///
/// **Whole matches only.** Every pattern is fenced by word boundaries, so a rule for
/// "cloud code" can never touch "Cloudflare" or the ordinary word "cloud".
///
/// **Glued words still match.** Engines run words together — "CloudCode", "cloud-code" — so
/// the gap between the parts of a phrase is matched as *optional* whitespace or hyphens
/// rather than a literal space.
public struct DictionaryCorrector: Sendable {
    private let rules: [Rule]
    private let phonetics: [String: String]

    private struct Rule: Sendable {
        let regex: NSRegularExpression
        let replacement: String
        let trigger: String
    }

    /// Longest span the phonetic pass will consider collapsing into one entry.
    ///
    /// Three, because a mishearing splits a name into more words than it started with —
    /// "Anthropic" arrives as "and thropic" — but past three the window starts swallowing
    /// ordinary sentences whole.
    public static let maxSpanWords = 3

    public init(entries: [DictionaryEntry]) {
        // Longest trigger first, so "Claude Code" wins over "Claude" and rewrites the span.
        let corrections = entries
            .filter { $0.isEnabled && $0.kind == .correction }
            .filter { !$0.hear.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .sorted { $0.hear.count > $1.hear.count }

        rules = corrections.compactMap { entry in
            guard let regex = Self.makeRegex(for: entry.hear) else { return nil }
            return Rule(
                regex: regex,
                replacement: NSRegularExpression.escapedTemplate(for: entry.write),
                trigger: entry.hear
            )
        }

        phonetics = Self.makePhonetics(from: entries)
    }

    public var isEmpty: Bool { rules.isEmpty && phonetics.isEmpty }

    /// Applies every rule in order, then the phonetic pass.
    ///
    /// - Returns: the rewritten text, plus one `AppliedCorrection` per distinct replacement
    ///   that fired.
    public func apply(to text: String) -> (text: String, applied: [AppliedCorrection]) {
        guard !isEmpty, !text.isEmpty else { return (text, []) }

        // NFC on both sides, or an accented trigger silently never matches. See AGENTS.md.
        var result = text.precomposedStringWithCanonicalMapping
        var applied: [AppliedCorrection] = []

        for rule in rules {
            let range = NSRange(result.startIndex..., in: result)
            let matches = rule.regex.numberOfMatches(in: result, range: range)
            guard matches > 0 else { continue }

            // Record what the engine produced, not the trigger: the real mishearing is the point.
            let firstMatch = rule.regex.firstMatch(in: result, range: range)
            let heard = firstMatch
                .flatMap { Range($0.range, in: result) }
                .map { String(result[$0]) } ?? rule.trigger

            result = rule.regex.stringByReplacingMatches(
                in: result,
                range: range,
                withTemplate: rule.replacement
            )

            Self.record(
                &applied,
                from: heard,
                to: rule.replacement.replacingOccurrences(of: "\\", with: ""),
                count: matches
            )
        }

        result = applyPhonetics(to: result, applied: &applied)

        return (result, applied)
    }

    // MARK: - Phonetic pass

    /// The sound-alike index: phonetic code to the text that should be written instead.
    ///
    /// An entry earns a place only if `PhoneticKey.isDistinctive` clears the phrase it would
    /// match on — the word itself for a term, the trigger for a correction. Naming a trigger
    /// yourself is not enough to make it safe: a rule reading "I in" to "Ayen" also rewrites
    /// "I am", because those are the same sound. A rejected entry keeps its literal rule and
    /// loses only the sound-alike family.
    ///
    /// First entry in file order wins a contested code, so the result is stable.
    private static func makePhonetics(from entries: [DictionaryEntry]) -> [String: String] {
        var index: [String: String] = [:]

        for entry in entries where entry.isEnabled {
            let write = entry.write.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !write.isEmpty else { continue }

            let source = entry.kind == .term
                ? write
                : entry.hear.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !source.isEmpty, PhoneticKey.isDistinctive(source) else { continue }

            let key = PhoneticKey.encode(source)
            guard index[key] == nil else { continue }
            index[key] = write
        }

        return index
    }

    /// Rewrites every span that sounds like something in the index.
    ///
    /// Longest span wins and spans never overlap, so a three-word match consumes its words and
    /// the scan resumes after them rather than re-matching a sub-span.
    private func applyPhonetics(to text: String, applied: inout [AppliedCorrection]) -> String {
        guard !phonetics.isEmpty else { return text }

        let words = Self.words(in: text)
        guard !words.isEmpty else { return text }

        var edits: [(range: Range<String.Index>, heard: String, replacement: String)] = []
        var start = 0

        while start < words.count {
            var span = min(Self.maxSpanWords, words.count - start)
            var matched = false

            while span >= 1 {
                if Self.isJoinable(words, from: start, count: span, in: text) {
                    let range = words[start].lowerBound..<words[start + span - 1].upperBound
                    let heard = String(text[range])
                    let key = PhoneticKey.encode(heard)

                    if key.count > 1, let replacement = phonetics[key],
                       heard.compare(replacement, options: .caseInsensitive) != .orderedSame {
                        edits.append((range, heard, replacement))
                        start += span
                        matched = true
                        break
                    }
                }
                span -= 1
            }

            if !matched { start += 1 }
        }

        guard !edits.isEmpty else { return text }

        var result = text
        for edit in edits.reversed() {
            result.replaceSubrange(edit.range, with: edit.replacement)
        }
        for edit in edits {
            Self.record(&applied, from: edit.heard, to: edit.replacement, count: 1)
        }

        return result
    }

    /// The ranges of every word in `text`, in order.
    private static func words(in text: String) -> [Range<String.Index>] {
        guard let regex = try? NSRegularExpression(pattern: "[\\p{L}\\p{N}']+") else { return [] }
        let range = NSRange(text.startIndex..., in: text)
        return regex.matches(in: text, range: range).compactMap { Range($0.range, in: text) }
    }

    /// Whether the words `start..<start + count` are separated only by spaces or hyphens.
    ///
    /// Punctuation between two words means they belong to different thoughts, and a span that
    /// reaches across a full stop would let the pass rewrite the seam between two sentences.
    private static func isJoinable(
        _ words: [Range<String.Index>],
        from start: Int,
        count: Int,
        in text: String
    ) -> Bool {
        for offset in 1..<max(count, 1) {
            let gap = text[words[start + offset - 1].upperBound..<words[start + offset].lowerBound]
            if gap.contains(where: { !$0.isWhitespace && $0 != "-" }) { return false }
        }
        return true
    }

    /// Adds a correction, folding it into an existing record for the same replacement.
    ///
    /// One record per replacement, because that is what the shared vectors contract promises
    /// and what history renders as a single badge.
    private static func record(
        _ applied: inout [AppliedCorrection],
        from: String,
        to: String,
        count: Int
    ) {
        guard let existing = applied.firstIndex(where: { $0.to == to }) else {
            applied.append(AppliedCorrection(from: from, to: to, count: count))
            return
        }
        let previous = applied[existing]
        applied[existing] = AppliedCorrection(
            from: previous.from,
            to: to,
            count: previous.count + count
        )
    }

    // MARK: - Literal rules

    /// Builds the pattern for one trigger phrase.
    ///
    /// The parts are joined with `[\s\-]*` — zero or more spaces or hyphens — which is what
    /// catches "CloudCode" and "Cloud-Code" alongside the spaced form.
    ///
    /// The fences are lookarounds on letters and digits rather than `\b`. `\b` would treat a
    /// trailing hyphen or apostrophe as a boundary and let a rule bite into a longer word;
    /// requiring that no letter or digit sits on either side is the stricter guarantee, and
    /// it's what keeps "cloud code" off "Cloudflare".
    private static func makeRegex(for trigger: String) -> NSRegularExpression? {
        // NFC here too: a trigger typed in the UI and one read from disk differ in normal form.
        let parts = trigger
            .precomposedStringWithCanonicalMapping
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .split(whereSeparator: { $0 == " " || $0 == "-" || $0 == "\t" })
            .map { NSRegularExpression.escapedPattern(for: String($0)) }

        guard !parts.isEmpty else { return nil }

        let body = parts.joined(separator: "[\\s\\-]*")
        let pattern = "(?<![\\p{L}\\p{N}])\(body)(?![\\p{L}\\p{N}])"

        return try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive])
    }
}

// MARK: - Engine biasing

public extension DictionaryCorrector {
    /// The phrases to hand the speech engine as context before it transcribes.
    ///
    /// Kept deliberately short. These models drift when given a long context list — on quiet
    /// or ambiguous audio they start inventing text from the vocabulary they were primed
    /// with, which is a far worse failure than the misspelling it was meant to fix.
    static let biasLimit = 40

    /// - Returns: the correct spellings — `.term` words and the *write* side of corrections —
    ///   most recently useful first, capped at `biasLimit`.
    static func biasPhrases(from entries: [DictionaryEntry]) -> [String] {
        var seen = Set<String>()
        var phrases: [String] = []

        for entry in entries where entry.isEnabled {
            let phrase = entry.write.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !phrase.isEmpty, seen.insert(phrase.lowercased()).inserted else { continue }
            phrases.append(phrase)
            if phrases.count == biasLimit { break }
        }

        return phrases
    }
}

File: windows/src/Murmur.Dictionary/DictionaryCorrector.cs
using System.Text;
using System.Text.RegularExpressions;

namespace Murmur.Dictionary;

/// <summary>One correction that actually fired.</summary>
public sealed record AppliedCorrection(string From, string To, int Count);

/// <summary>
/// Rewrites transcribed text using the dictionary's correction pairs.
/// </summary>
/// <remarks>
/// <para>
/// This is the guaranteed half of the dictionary. Engine biasing is a nudge — it raises the
/// odds of the right word and promises nothing — so anything that must be correct is fixed
/// here, after the fact, deterministically.
/// </para>
/// <para>
/// A direct counterpart to <c>DictionaryCorrector.swift</c>. The two are independent
/// implementations of one contract, and <c>shared/dictionary-test-vectors.json</c> is what
/// stops them drifting. <b>Change the semantics there first.</b>
/// </para>
/// <para>Three rules, all load-bearing:</para>
/// <list type="number">
/// <item><b>Longest match first.</b> "Claude Code" is applied before "Claude", so the longer
/// rule isn't pre-empted by a shorter one that overlaps it.</item>
/// <item><b>Whole matches only.</b> Every pattern is fenced by word boundaries, so a rule for
/// "cloud code" can never touch "Cloudflare" or the ordinary word "cloud".</item>
/// <item><b>Glued words still match.</b> Engines run words together — "CloudCode",
/// "cloud-code" — so the gap between parts is matched as optional whitespace or hyphens.</item>
/// </list>
/// <para>
/// A second, phonetic pass then rewrites spans that merely <i>sound</i> like something the
/// dictionary knows, so one entry covers every way an engine can fumble a name.
/// </para>
/// </remarks>
public sealed class DictionaryCorrector
{
    /// <summary>
    /// Guards against a pathological dictionary hanging a dictation. Matching is linear here,
    /// so this should never trigger; it exists so that a bug can't wedge the app.
    /// </summary>
    private static readonly TimeSpan MatchTimeout = TimeSpan.FromSeconds(1);

    private static readonly char[] PhraseSeparators = [' ', '-', '\t'];

    private static readonly Regex WordPattern =
        new(@"[\p{L}\p{N}']+", RegexOptions.CultureInvariant, MatchTimeout);

    /// <summary>
    /// Longest span the phonetic pass will consider collapsing into one entry.
    /// </summary>
    /// <remarks>
    /// Three, because a mishearing splits a name into more words than it started with —
    /// "Anthropic" arrives as "and thropic" — but past three the window starts swallowing
    /// ordinary sentences whole.
    /// </remarks>
    public const int MaxSpanWords = 3;

    private readonly List<Rule> _rules;
    private readonly Dictionary<string, string> _phonetics;

    private sealed record Rule(Regex Regex, string Replacement, string Trigger);

    /// <summary>Compiles the enabled correction entries into an ordered rule set.</summary>
    /// <param name="entries">The dictionary. Terms and disabled entries are ignored here.</param>
    public DictionaryCorrector(IEnumerable<DictionaryEntry> entries)
    {
        // Longest trigger first; the stable sort keeps parity with Swift. See AGENTS.md.
        _rules = entries
            .Where(e => e.IsEnabled && e.Kind == EntryKind.Correction)
            .Where(e => !string.IsNullOrWhiteSpace(e.Hear))
            .OrderByDescending(e => e.Hear.Length)
            .Select(e => MakeRule(e.Hear, e.Write))
            .OfType<Rule>()
            .ToList();

        _phonetics = MakePhonetics(entries);
    }

    /// <summary>True when no enabled entry produced a literal rule or a sound-alike code.</summary>
    public bool IsEmpty => _rules.Count == 0 && _phonetics.Count == 0;

    /// <summary>Applies every rule in order.</summary>
    /// <param name="text">Raw transcribed text.</param>
    /// <returns>The rewritten text, plus one entry per rule that fired.</returns>
    public (string Text, IReadOnlyList<AppliedCorrection> Applied) Apply(string text)
    {
        if (IsEmpty || string.IsNullOrEmpty(text)) return (text, []);

        // NFC before matching, exactly as the Swift side does. See AGENTS.md.
        var result = text.Normalize(NormalizationForm.FormC);
        var applied = new List<AppliedCorrection>();

        foreach (var rule in _rules)
        {
            var matches = rule.Regex.Matches(result);
            if (matches.Count == 0) continue;

            var heard = matches[0].Value;

            // MatchEvaluator keeps the replacement literal. See AGENTS.md.
            result = rule.Regex.Replace(result, _ => rule.Replacement);

            Record(applied, heard, rule.Replacement, matches.Count);
        }

        result = ApplyPhonetics(result, applied);

        return (result, applied);
    }

    /// <summary>
    /// The sound-alike index: phonetic code to the text that should be written instead.
    /// </summary>
    /// <remarks>
    /// An entry earns a place only if <see cref="PhoneticKey.IsDistinctive"/> clears the phrase
    /// it would match on — the word itself for a term, the trigger for a correction. Naming a
    /// trigger yourself is not enough to make it safe: a rule reading "I in" to "Ayen" also
    /// rewrites "I am", because those are the same sound. A rejected entry keeps its literal
    /// rule and loses only the sound-alike family.
    /// <para>First entry in file order wins a contested code, so the result is stable.</para>
    /// </remarks>
    private static Dictionary<string, string> MakePhonetics(IEnumerable<DictionaryEntry> entries)
    {
        var index = new Dictionary<string, string>(StringComparer.Ordinal);

        foreach (var entry in entries.Where(e => e.IsEnabled))
        {
            var write = entry.Write.Trim();
            if (write.Length == 0) continue;

            var source = entry.Kind == EntryKind.Term ? write : entry.Hear.Trim();
            if (source.Length == 0 || !PhoneticKey.IsDistinctive(source)) continue;

            var key = PhoneticKey.Encode(source);
            if (!index.ContainsKey(key)) index[key] = write;
        }

        return index;
    }

    /// <summary>Rewrites every span that sounds like something in the index.</summary>
    /// <remarks>
    /// Longest span wins and spans never overlap, so a three-word match consumes its words and
    /// the scan resumes after them rather than re-matching a sub-span.
    /// </remarks>
    private string ApplyPhonetics(string text, List<AppliedCorrection> applied)
    {
        if (_phonetics.Count == 0) return text;

        var words = WordPattern.Matches(text);
        if (words.Count == 0) return text;

        var edits = new List<(int Start, int Length, string Heard, string Replacement)>();
        var start = 0;

        while (start < words.Count)
        {
            var span = Math.Min(MaxSpanWords, words.Count - start);
            var matched = false;

            while (span >= 1)
            {
                if (IsJoinable(words, start, span, text))
                {
                    var begin = words[start].Index;
                    var length = words[start + span - 1].Index + words[start + span - 1].Length - begin;
                    var heard = text[begin..(begin + length)];
                    var key = PhoneticKey.Encode(heard);

                    if (key.Length > 1
                        && _phonetics.TryGetValue(key, out var replacement)
                        && !string.Equals(heard, replacement, StringComparison.OrdinalIgnoreCase))
                    {
                        edits.Add((begin, length, heard, replacement));
                        start += span;
                        matched = true;
                        break;
                    }
                }

                span--;
            }

            if (!matched) start++;
        }

        if (edits.Count == 0) return text;

        var builder = new StringBuilder(text);
        for (var i = edits.Count - 1; i >= 0; i--)
        {
            builder.Remove(edits[i].Start, edits[i].Length).Insert(edits[i].Start, edits[i].Replacement);
        }

        foreach (var edit in edits) Record(applied, edit.Heard, edit.Replacement, 1);

        return builder.ToString();
    }

    /// <summary>
    /// Whether the words <c>start</c> to <c>start + count</c> are separated only by spaces or
    /// hyphens.
    /// </summary>
    /// <remarks>
    /// Punctuation between two words means they belong to different thoughts, and a span that
    /// reaches across a full stop would let the pass rewrite the seam between two sentences.
    /// </remarks>
    private static bool IsJoinable(MatchCollection words, int start, int count, string text)
    {
        for (var offset = 1; offset < count; offset++)
        {
            var previous = words[start + offset - 1];
            var gapStart = previous.Index + previous.Length;
            var gap = text.AsSpan(gapStart, words[start + offset].Index - gapStart);

            foreach (var c in gap)
            {
                if (!char.IsWhiteSpace(c) && c != '-') return false;
            }
        }

        return true;
    }

    /// <summary>Adds a correction, folding it into an existing record for the same replacement.</summary>
    /// <remarks>
    /// One record per replacement, because that is what the shared vectors contract promises
    /// and what history renders as a single badge.
    /// </remarks>
    private static void Record(List<AppliedCorrection> applied, string from, string to, int count)
    {
        var existing = applied.FindIndex(a => string.Equals(a.To, to, StringComparison.Ordinal));
        if (existing < 0)
        {
            applied.Add(new AppliedCorrection(from, to, count));
            return;
        }

        applied[existing] = applied[existing] with { Count = applied[existing].Count + count };
    }

    /// <summary>Builds the pattern for one trigger phrase.</summary>
    /// <remarks>
    /// <para>
    /// Parts are joined with <c>[\s\-]*</c> — zero or more spaces or hyphens — which catches
    /// "CloudCode" and "Cloud-Code" alongside the spaced form.
    /// </para>
    /// <para>
    /// The fences are lookarounds on letters and digits rather than <c>\b</c>. <c>\b</c>
    /// treats a trailing hyphen or apostrophe as a boundary and would let a rule bite into a
    /// longer word; requiring that no letter or digit sits on either side is the stricter
    /// guarantee, and it's what keeps "cloud code" off "Cloudflare".
    /// </para>
    /// </remarks>
    private static Rule? MakeRule(string trigger, string replacement)
    {
        var parts = trigger
            .Normalize(NormalizationForm.FormC)
            .Trim()
            .Split(PhraseSeparators, StringSplitOptions.RemoveEmptyEntries)
            .Select(Regex.Escape)
            .ToArray();

        if (parts.Length == 0) return null;

        var body = string.Join(@"[\s\-]*", parts);
        var pattern = $@"(?<![\p{{L}}\p{{N}}]){body}(?![\p{{L}}\p{{N}}])";

        try
        {
            var regex = new Regex(pattern, RegexOptions.IgnoreCase | RegexOptions.CultureInvariant, MatchTimeout);
            return new Rule(regex, replacement, trigger);
        }
        catch (ArgumentException)
        {
            return null;
        }
    }


    /// <summary>
    /// How many phrases to hand the speech engine as context.
    /// </summary>
    /// <remarks>
    /// Deliberately small. These models drift when given a long context list — on quiet or
    /// ambiguous audio they start inventing text from the vocabulary they were primed with,
    /// which is a far worse failure than the misspelling it was meant to fix.
    /// </remarks>
    public const int BiasLimit = 40;

    /// <summary>
    /// The correct spellings — Term words and the <i>write</i> side of corrections — capped
    /// at <see cref="BiasLimit"/>, de-duplicated case-insensitively, in file order.
    /// </summary>
    public static IReadOnlyList<string> BiasPhrases(IEnumerable<DictionaryEntry> entries)
    {
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var phrases = new List<string>();

        foreach (var entry in entries.Where(e => e.IsEnabled))
        {
            var phrase = entry.Write.Trim();
            if (phrase.Length == 0 || !seen.Add(phrase)) continue;
            phrases.Add(phrase);
            if (phrases.Count == BiasLimit) break;
        }

        return phrases;
    }
}
